mlflow/models/signature.py

```diff
@@ -0,0 +1,449 @@
+            raise TypeError("inputs must be mlflow.models.signature.Schema, got '{}'".format(
+                type(inputs)))
+        if outputs is not None and not isinstance(outputs, Schema):
+            raise TypeError("outputs must be either None or mlflow.models.signature.Schema, "
+                            "got '{}'".format(type(inputs)))
+        self.inputs = inputs
+        self.outputs = outputs
+
+    def to_dict(self) -> Dict[str, str]:
+        """
+        Serialize into a 'jsonable' dictionary.
+
+        Input and output schema are represented as json strings. This is so that the
+        representation is compact when embedded in a MLmofel yaml file.
+
+        :return: dictionary representation with input and output shcema represented as json strings.
+        """
+
+        return {
+            "inputs": self.inputs.to_json(),
+            "outputs": self.outputs.to_json() if self.outputs is not None else None
+        }
+
+    @classmethod
+    def from_dict(cls, signature_dict: Dict[str, str]):
+        """
+        Deserialize from dictionary representation.
+
+        :param signature_dict: Dictionary representation of model signature.
+                               Expected dictionary format:
+                               `{'inputs': <json string>, 'outputs': <json string>" }`
+
+        :return: ModelSignature populated with the data form the dictionary.
+        """
+        inputs = Schema.from_json(signature_dict["inputs"])
+        if "outputs" in signature_dict and signature_dict["outputs"] is not None:
+            outputs = Schema.from_json(signature_dict["outputs"])
+            return cls(inputs, outputs)
+        else:
+            return cls(inputs)
+
+    def __eq__(self, other) -> bool:
+        return self.inputs == other.inputs and self.outputs == other.outputs
+
+    def __repr__(self) -> str:
+        return json.dumps({"ModelSignature": self.to_dict()}, indent=2)
+
+
+ModelInputExample = TypeVar('ModelInputExample', pd.DataFrame, np.ndarray, dict, list)
+MlflowModelDataset = TypeVar('MlflowModelDataset', pd.DataFrame, np.ndarray, Dict[str, np.ndarray])
+
+
+def infer_signature(model_input: MlflowModelDataset,
+                    model_output: MlflowModelDataset = None) -> ModelSignature:
+    """
+    Infer an MLflow model signature from the training data (input) and model predictions (output).
+    This method captures the column names and data types from the user data. The signature
+    represents model input and output as dataframes with (optionally) named columns and data type
+    specified as one of types defined in :py:class:`DataType`. This method will raise
+    an exception if the user data contains incompatible types or is not passed in one of the
+    supported formats (containers).
+
+    Supported input:
+
+    container types:
+      - pandas.DataFrame
+      - dictionary of { name -> numpy.ndarray}
+      - numpy.ndarray
+      - pyspark.sql.DataFrame
```

Comment:
Do we actually support this one? seems like the TypeVar does not include it.

Reply:
We do support it. The typevar can not have it because it requires importing spark. The typevar is more for a hint, the types are not enforced. I guess alternatively we could mode infer)signature with spark frame into mfllow.spark package but I think this one is nicer, despite the missing type. Maybe I can make the typevar conditional depending on whether pyspark is available.